File: leetcode/first-missing-positive.py
# 41. First Missing Positive
# 🔴 Hard
#
# https://leetcode.com/problems/first-missing-positive/
#
# Tags: Array - Hash Table

import timeit
from typing import List


# Use the pigeon hole principle with the array positions as the holes.
# We visit each position on the array, if we find a positive integer
# that would fit a "hole", that is, one of the array positions between
# 0 and len(arr), we swap it with the value there, we keep doing this
# until the current index has a value that does not belong in any of
# the holes, or the destination hole already has its correct value, then
# we move to the next hole. Once we iterate over the entire array, we
# know that holes that have their corresponding value in the array will
# be filled with the correct value, we iterate over the array from the
# start, the index of the first hole that contains the wrong pigeon is
# the first missing positive.
#
# Time complexity: O(n) - Since on each operation we are placing, at
# least, one of the values in its correct pigeon hole, we will do that
# operation at most n-1 times, then we will skip these correct positions
# in the outer loop, even with the nested for-while loop on the first
# pass, the complexity remains O(n)
# Space complexity: O(1) - Constant space.
#
# Runtime: 770 ms, faster than 40.67%
# Memory Usage: 27.2 MB, less than 82.84%
class PigeonHoling:
    def firstMissingPositive(self, nums: List[int]) -> int:
        # Place values i in the 1..len(nums)-1 at the nums[i-1]
        # index, ignore other values.
        for i in range(len(nums)):
            # Keep swapping the value while this position has a value
            # that could be placed in a hole. Skip if the hole already
            while (
                nums[i] > 0
                and nums[i] <= len(nums)
                and nums[i] != i + 1
                and nums[i] != nums[nums[i] - 1]
            ):
                # Store the destination index in a variable.
                dest_idx = nums[i] - 1
                # Swap this value with the value at nums[nums[i]-1]
                nums[i], nums[dest_idx] = nums[dest_idx], nums[i]
                # nums[i], nums[nums[i] - 1] = nums[nums[i] - 1], nums[i]
        # Iterate over the modified array to find the first missing positive.
        i = 0
        while i < len(nums):
            if nums[i] != i + 1:
                break
            i += 1
        return i + 1


def test():
    executors = [PigeonHoling]
    tests = [
        [[1, 1], 2],
        [[1, 2, 0], 3],
        [[3, 4, -1, 1], 2],
        [[7, 8, 9, 11, 12], 1],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.firstMissingPositive(t[0])
                exp = t[1]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/first-missing-positive.rs
// 41. First Missing Positive
// 🔴 Hard
//
// https://leetcode.com/problems/first-missing-positive/
//
// Tags: Array - Hash Table

struct Solution;
impl Solution {
    /// The input constraints mean that, either the input vector contains exactly the first n
    /// integers between 1..=n or one or more of them are missing, either way, we do not care
    /// about any values except one instance each of these integers, anything else we can ignore,
    /// for example duplicates, negative numbers, zero and anything bigger than n. We iterate over
    /// the input, each time we find a value x in the range 1..=n we move it to the index x-1,
    /// unless the value there is already x, once we complete that loop, each index i will contain
    /// the integer i+1 if it was found in the original input, we iterate over the mutated input
    /// vector and return the first nums[i] where the value is not i+1, if we get to the end of the
    /// vector without having returned anything, it means that it contains exactly the values 1..=n
    /// and we know that the first missing positive is n+1, we can return that value.
    ///
    /// Time complexity: O(n) - Even though we have a nested while loop inside the for loop, all it
    /// does is move a positive integer in the valid range to its position by index, each value
    /// is moved at most once so the inner loop runs n times, it could run multiple times for one
    /// iteration of the outer for, but that would mean that it does not need to run in the same
    /// number of other iterations of the for loop.
    /// Space complexity: O(1) - We use constant extra memory by using the input vector to mark
    /// elements seen.
    ///
    /// Runtime 6 ms Beats 63%
    /// Memory 2.88 MB Beats 99%
    pub fn first_missing_positive(mut nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut num;
        for i in 0..n {
            num = nums[i];
            while num > 0
                && num <= n as i32
                && num as usize - 1 != i
                && nums[i] != nums[num as usize - 1]
            {
                nums.swap(i, num as usize - 1);
                num = nums[i];
            }
        }
        for i in 0..n {
            if nums[i] != i as i32 + 1 {
                return i as i32 + 1;
            }
        }
        n as i32 + 1
    }
}

// Tests.
fn main() {
    let tests = [
        (vec![1, 1, 1], 2),
        (vec![1, 2, 0], 3),
        (vec![1, 2, 3], 4),
        (vec![3, 4, -1, 1], 2),
        (vec![-3, -4, -1, 1], 2),
        (vec![7, 8, 9, 11, 12], 1),
    ];
    println!("\n\x1b[92m» Running {} tests...\x1b[0m", tests.len());
    let mut success = 0;
    for (i, t) in tests.iter().enumerate() {
        let res = Solution::first_missing_positive(t.0.clone());
        if res == t.1 {
            success += 1;
            println!("\x1b[92m✔\x1b[95m Test {} passed!\x1b[0m", i);
        } else {
            println!(
                "\x1b[31mx\x1b[95m Test {} failed expected: {:?} but got {}!!\x1b[0m",
                i, t.1, res
            );
        }
    }
    println!();
    if success == tests.len() {
        println!("\x1b[30;42m✔ All tests passed!\x1b[0m")
    } else if success == 0 {
        println!("\x1b[31mx \x1b[41;37mAll tests failed!\x1b[0m")
    } else {
        println!(
            "\x1b[31mx\x1b[95m {} tests failed!\x1b[0m",
            tests.len() - success
        )
    }
}
